lib/cchain/provider/abci.go

```diff
@@ -763,3 +671,134 @@ func attsFromAtHeight(ctx context.Context, cl atypes.QueryClient, chainVer xchai
 
 	return atts, true, nil
 }
+
+// binarySearch uses a binary search between defined start and end consensus block heights to
+// find the attestations with the provided fromOffset.
+func binarySearch(
+	ctx context.Context,
+	attCl atypes.QueryClient,
+	chainVer xchain.ChainVersion,
+	chainName string,
+	fromOffset uint64,
+	startHeight uint64,
+	endHeight uint64,
+) (uint64, error) {
+	const endpoint = "search_offset"
+
+	for steps := 0; startHeight <= endHeight; steps++ {
+		midHeight := startHeight + umath.SubtractOrZero(startHeight, startHeight)/2
+
+		earliestAtt, ok, err := queryEarliestAttestation(ctx, attCl, chainVer, midHeight)
+		if err != nil {
+			incQueryErr(endpoint)
+			return 0, errors.Wrap(err, "abci query earliest-attestation-in-state")
+		}
+		if !ok {
+			// If we're so far back that there's no attestation at all, move forward
+			startHeight = midHeight + 1
+			continue
+		}
+
+		latestAtt, ok, err := queryLatestAttestation(ctx, attCl, chainVer, midHeight)
+		if err != nil {
+			incQueryErr(endpoint)
+			return 0, errors.Wrap(err, "abci query latest-attestation")
+		}
+
+		if !ok {
+			return 0, errors.New("no latest attestation found despite earlier check [BUG]")
+		}
+
+		if fromOffset >= earliestAtt.AttestOffset && fromOffset <= latestAtt.AttestOffset {
+			binarySearchStepsMetric(chainName, steps)
+
+			return midHeight, nil
+		}
+
+		// Query at a lower or higher height depending on whether fromOffset
+		// is smaller or larger than the earliest offset we found
+		if fromOffset < earliestAtt.AttestOffset {
+			endHeight = umath.SubtractOrZero(midHeight, 1)
+		} else {
+			startHeight = midHeight + 1
+		}
+	}
+
+	return 0, errors.New("unexpectedly reach end of search method [BUG]")
+}
+
+// lookbackRange does an exponential lookback from the latest block height provided
+// until attestation found has lower offset than the provided from offset, this
+// guarantees the fromOffset attestation will be found in the range.
+//
+//nolint:nonamedreturns // named returned for clarity
+func lookbackRange(
+	ctx context.Context,
+	attCl atypes.QueryClient,
+	chainVer xchain.ChainVersion,
+	chainName string,
+	fromOffset uint64,
+	latestHeight uint64,
+) (start uint64, end uint64, err error) {
+	const endpoint = "search_offset"
```

Comment:
copy paste, lets improve this label, also add latency metrics for this.

Reply:
I intentionally left them that way to not break existing counters for the same operation. I will change them if you think it's better.